packages/remix-server-runtime/serialize.ts

```diff
@@ -1,16 +1,41 @@
 import type { Jsonify } from "./jsonify";
 import type { TypedDeferredData, TypedResponse } from "./responses";
+import type {
+  ClientActionFunctionArgs,
+  ClientLoaderFunctionArgs,
+} from "./routeModules";
 import { expectType } from "./typecheck";
 import { type Expect, type Equal } from "./typecheck";
 
+// prettier-ignore
+/**
+ * Infer JSON serialized data type returned by a loader or action, while avoiding
+ * deserialization if the input type if a clientLoader/clientAction that returns
+ * a non-Response
+ *
+ * For example:
+ * `type LoaderData = SerializeFrom<typeof loader>`
+ */
+export type SerializeFrom<T> = (
+  T extends (...args: any[]) => infer Output
+    ? Output extends Promise<TypedResponse>
+      ? SerializeFromImpl<T>
+      : Parameters<T> extends [ClientLoaderFunctionArgs | ClientActionFunctionArgs]
+        ? Output extends Promise<TypedDeferredData<infer D>>
+          ? D // client defer() is just the raw data
+          : Awaited<Output> // client naked object is just the object
+        : SerializeFromImpl<T>
+    : SerializeFromImpl<T>
+);
```

Comment:
Exporting this as part of `SerializeFrom` ensures we get this across the board in `useLoaderData`, `useActionData`, `useFetcher`, `useRouteLoaderData`, meta `matches.data`, etc.